internal/provisioner/registry/controllers/bundle_controller.go

```diff
@@ -0,0 +1,285 @@
+//+kubebuilder:rbac:groups=authorization.k8s.io,resources=subjectaccessreviews,verbs=create
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *BundleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reconcileErr error) {
+	l := log.FromContext(ctx)
+	l.V(1).Info("starting reconciliation")
+	defer l.V(1).Info("ending reconciliation")
+	bundle := &rukpakv1alpha1.Bundle{}
+	if err := r.Get(ctx, req.NamespacedName, bundle); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	u := updater.NewBundleUpdater(r.Client)
+	defer func() {
+		if err := u.Apply(ctx, bundle); err != nil {
+			l.Error(err, "failed to update status")
+		}
+	}()
+	u.UpdateStatus(updater.EnsureObservedGeneration(bundle.Generation))
+
+	finalizerResult, err := r.Finalizers.Finalize(ctx, bundle)
+	if err != nil {
+		u.UpdateStatus(
+			updater.EnsureResolvedSource(nil),
+			updater.EnsureContentURL(""),
+			updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+			updater.EnsureCondition(metav1.Condition{
+				Type:    rukpakv1alpha1.TypeUnpacked,
+				Status:  metav1.ConditionUnknown,
+				Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+				Message: err.Error(),
+			}),
+		)
+		return ctrl.Result{}, err
+	}
+	var (
+		finalizerUpdateErrs []error
+	)
+	// Update the status subresource before updating the main object. This is
+	// necessary because, in many cases, the main object update will remove the
+	// finalizer, which will cause the core Kubernetes deletion logic to
+	// complete. Therefore, we need to make the status update prior to the main
+	// object update to ensure that the status update can be processed before
+	// a potential deletion.
+	if finalizerResult.StatusUpdated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Status().Update(ctx, bundle))
+	}
+	if finalizerResult.Updated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Update(ctx, bundle))
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bundle.GetDeletionTimestamp().IsZero() {
+		err := apimacherrors.NewAggregate(finalizerUpdateErrs)
+		if err != nil {
+			u.UpdateStatus(
+				updater.EnsureResolvedSource(nil),
+				updater.EnsureContentURL(""),
+				updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeUnpacked,
+					Status:  metav1.ConditionUnknown,
+					Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+					Message: err.Error(),
+				}),
+			)
+		}
+		return ctrl.Result{}, err
+	}
+
+	unpackResult, err := r.Unpacker.Unpack(ctx, bundle)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("source bundle content: %w", err))
+	}
+	switch unpackResult.State {
```

Comment:
Do we need to account for a failing source state here?

Reply:
No, the valid states are `Pending`, `Unpacking`, and `Unpacked`. A failing state would be just returning an error from `Unpack`.